### PR TITLE
FOGL-7277 Restructure OMF plugin such that the OMF object can be reused

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

File: C/plugins/north/OMF/include/omf.h [hunk 1]
```diff
@@ -514,6 +514,11 @@ class OMF
 		 * Service name
 		 */
 		const std::string	m_name;
+
+		/**
+		 * Has base types been sent to the PI Server
```
Comment (RVerhoeff) [inline C/plugins/north/OMF/include/omf.h:519]: _Have_ base types been sent to the PI Server

**Review by pintomax — DISMISSED** [review]

File: C/plugins/north/OMF/include/omfinfo.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		string		m_formatNumber;	        // OMF protocol Number format
+		string		m_formatInteger;	        // OMF protocol Integer format
+		OMF_ENDPOINT	m_PIServerEndpoint;     // Defines which End point should be used for the communication
+		NAMINGSCHEME_ENDPOINT
+				m_NamingScheme; // Define how the object names should be generated - https://fledge-iot.readthedocs.io/en/latest/OMF.html#naming-scheme
+		string		m_DefaultAFLocation;    // 1st hierarchy in Asset Framework, PI Web API only.
+		string		m_AFMap;                // Defines a set of rules to address where assets should be placed in the AF hierarchy.
+						//    https://fledge-iot.readthedocs.io/en/latest/OMF.html#asset-framework-hierarchy-rules
+
+		string		m_prefixAFAsset;        // Prefix to generate unique asste id
```
Comment (RVerhoeff) [inline C/plugins/north/OMF/include/omfinfo.h:151]: Typo: asste -> asset

File: C/plugins/north/OMF/plugin.cpp [hunk 3]
```diff
@@ -812,89 +416,9 @@ void plugin_start(const PLUGIN_HANDLE handle,
...
-	}
-	else
-	{
-		Logger::getLogger()->info("%s plugin is using per asset OMF prefix %s=%d "
-					  "(max value found)",
-					  PLUGIN_NAME,
-					  TYPE_ID_KEY,
-					  getMaxTypeId(connInfo));
-	}
+	OMFInformation *info = (OMFInformation *)handle;
```
Comment (RVerhoeff) [inline C/plugins/north/OMF/plugin.cpp:419]: We had an outside developer a while ago who recommended coding this cast as
`OMFInformation *info = const_cast<OMFInformation *>(handle);`
Certainly not a major problem, or even required. What do you think?

File: C/plugins/north/OMF/include/omfinfo.h [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		std::string	AuthBasicCredentialsGenerate(std::string& userId, std::string& password);
+		void		AuthKerberosSetup(std::string& keytabEnv, std::string& keytabFileName);
+		double		GetElapsedTime(struct timeval *startTime);
+		bool		IsPIWebAPIConnected();
+		
+	private:
+		Logger		*m_logger;
+		HttpSender	*m_sender;              // HTTPS connection
+		OMF 		*m_omf;                 // OMF data protocol
+		bool		m_sendFullStructure;    // It sends the minimum OMF structural messages to load data into Data Archive if disabled
```
Comment (RVerhoeff) [inline C/plugins/north/OMF/include/omfinfo.h:132]: PI Data Archive

File: C/plugins/north/OMF/omf.cpp [hunk 5]
```diff
@@ -489,7 +490,8 @@ bool OMF::sendDataTypes(const Reading& row, OMFHints *hints)
 		return false;
 	}
 
-	if (m_sendFullStructure) {
+	if (m_sendFullStructure)
+	{
```
Comment (RVerhoeff) [inline C/plugins/north/OMF/omf.cpp:495]: Two blank lines are not needed

**Review by RVerhoeff — CHANGES_REQUESTED** [review]
I thought that in order to keep context between calls to _plugin_send_ that an _OMF_ object would be created once and would be reused in every send. Instead, you created the _OMFInformation_ object for that role. I am curious why using the _OMF_ object directly was not an option.

I built Fledge from this branch and re-ran the scenario in [FOGL-7837](https://scaledb.atlassian.net/browse/FOGL-[7837](https://scaledb.atlassian.net/browse/FOGL-7837)). It does indeed call the type and container creation endpoints once and then performs only Data POSTs for the timeseries data. Very nice!

I found typos in parts of the code not changed in this PR. Both are in _omf.cpp_:

- Line 452:
`Logger::getLogger()->warn("The OMF endpoint reported a bad request when sending dta type contianers : %d messages",`
Typos: dta -> data, contianers -> containers

- Line 4707:
`* @return true If the data types were sent correctly. Otherwsie false.`
Typo: Otherwsie -> Otherwise

**Comment by MarkRiddoch** [comment]
> 

It is done this way to make the minimum of changes as we wanted this in place quickly and with least risk. The old ConnectorInformation structure got made into a class and otherwise very little needed to be changed.

File: C/plugins/north/OMF/plugin.cpp (hunk 3 above)
Comment (MarkRiddoch) [reply C/plugins/north/OMF/plugin.cpp:419]: It really makes no difference

**Review by pintomax — APPROVED** [review]

**Review by RVerhoeff — APPROVED** [review]

**Pull request merged**